### ISSUE TITLE
[K8s Testing] - Test Executors on Sourcegraph Deployments

### ISSUE THREAD

**Issue by kalanchan** [sourcegraph/sourcegraph#50824]
# Sourcegraph Deployments compatibility

Ensure Executors for Kubernetes works on all Sourcegraph deployment methods

Goal: Test Executors on customer Sourcegraph deployments methods

**Note:** Some of the deployment methods we might not have test environments for, we might have to create a separate issue to create them and the deployment manifests as well. Please identify which ones we couldn't test on and we'll work on adding them in after

```[tasklist]
## Sourcegraph Deployments testing (as many environments as you can)
- [ ] Deploy Executors to [Base](https://github.com/sourcegraph/deploy-sourcegraph)
- [x] Deploy Executors to [Helm](https://github.com/sourcegraph/deploy-sourcegraph-helm) (k8s-dogfood)
- [x] Deploy Executors to [Kustomize](https://github.com/sourcegraph/deploy-sourcegraph-k8s)
- [x] Check Executor job spawns, CPU/Mem usage, Job completion & lifecycle
- [x] Monitor Cluster usage spikes, determine how jobs get throttled upon cluster resource limitations
```

### Batch Changes

Ensure Native SSBC is feature flag enabled.

1. Create a new Batch Change
2. Have 10+ repositories be selected by the batch spec
3. Run the batch change
4. Ensure the batch change is applied correctly to the workspace.

### Code Intel

1. Enable [Precise Code Nav](https://docs.sourcegraph.com/code_navigation/how-to/enable_auto_indexing).
2. run indexing across 10+ repositories.

**Comment by sanderginn** [comment]
After extensive testing, here are my findings.

### Observations
- Memory and cpu usage all seemed normal. All jobs that I inspected remained under 1 cpu and under 2Gi mem
- Permission issues. Fix (without tests) in https://github.com/sourcegraph/sourcegraph/pull/51418

### Improvements
- Metrics in Grafana all seem to be focused on VM deployments. Unless I missed them, we should probably add k8s metrics?
- Document deploying executors so that jobs always end up on the same node as the executor
- Explicitly document that `externalUrl` needs to be configured in the site configuration. It's not very likely that customers won't have this configured, but it's a pretty opaque issue to debug in the event that it is true. `codeintel` jobs will try to upload their results to dotcom when it is unset.

### Scheduling options
So, the current approach doesn't really accommodate for node autoscaling. The executor can currently only configure node selectors or affinities for the spawned jobs. This works fine if there's a single node that runs all executor jobs, as the node can be targeted through e.g. a label. However, if any additional nodes are added with the same label, there is no guarantee that a node selector or affinity is going to choose the same node as the executor pod any longer.  

Initially, I thought this could be resolved by configuring [pod affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity), but this actually runs into the same issue... you can configure affinity based on pod labels, but those labels are identical across all pods. In other words, this doesn't solve anything.  

**But**, I think there might be another solution. Using the [downward API](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/#downwardapi-fieldRef), we can use the `spec.nodeName` property of the executor pod when setting the job's node selector in the `EXECUTOR_KUBERNETES_NODE_NAME` environment variable:
```yaml
    - name: EXECUTOR_KUBERNETES_NODE_NAME
      valueFrom:
        fieldRef:
          fieldPath: spec.nodeName
```
This way, it doesn't matter how many replicas of an executor are running, because the specific replica will always generate jobs on the same node as it is running. Deploying the executor as a `DaemonSet`, optionally with node taints and tolerations configured, will ensure that a replica is running on all nodes that can run executor jobs. We would have to refactor the code to add node tolerations to the jobs spawned by the executor, though.

:tada:, right? Well, there's a caveat.. it might cause jobs to become unschedulable, at least for a long time. If a node is at capacity resource-wise, the executor can generate jobs all it wants, but those will be stuck in `Pending` until other jobs terminate and free up resources. I don't think this is a fundamental issue, because by design executor jobs are supposed to terminate, so any job that can't run yet should eventually be up for execution.. it just makes it harder to offer an out-of-the-box experience that works for most customers.  
For example, if a customer's cluster uses relatively small machine types to power its nodes, but it's running resource-heavy batch changes, it's going to be able to execute only a few jobs in parallel. Both the node configuration and the executor can be fine-tuned to perform better, but we'll have to do performance testing to determine those parameters and document them _thoroughly_.

### Unresolved errors
- Sometimes jobs could not successfully create the step container, producing a `CreateContainerError`:
   ```shell
   Events:
     Type     Reason     Age                   From               Message
     ----     ------     ----                  ----               -------
     Normal   Scheduled  9m2s                  default-scheduler  Successfully assigned sourcegraph/sg-executor-job-codeintel-352-step.kubernetes.pre-index.0-x72q9 to gke-organization-sourcegraph-executor-4bba78c7-6zkr
     Warning  Failed     4m20s (x9 over 6m7s)  kubelet            Error: failed to reserve container name "sg-executor-job-container_sg-executor-job-codeintel-352-step.kubernetes.pre-index.0-x72q9_sourcegraph_7ab8a38d-e78b-4dc9-854b-01050704b1fb_0": name "sg-executor-job-container_sg-executor-job-codeintel-352-step.kubernetes.pre-index.0-x72q9_sourcegraph_7ab8a38d-e78b-4dc9-854b-01050704b1fb_0" is reserved for "a3bb8963720308636d9ff324ac402ec6e04ab46a72c73cd2ab4a74d449243872"
     Warning  Failed     2m9s (x2 over 6m8s)   kubelet            Error: context deadline exceeded
     Normal   Pulled     102s (x13 over 8m8s)  kubelet            Container image "sourcegraph/lsif-go@sha256:2194d2652862966f022b537ed81bccf5a9a535ab763534cb4e98a3083c8a1bc6" already present on machine
     Warning  Failed     102s (x2 over 117s)   kubelet            Error: failed to reserve container name "sg-executor-job-container_sg-executor-job-codeintel-352-step.kubernetes.pre-index.0-x72q9_sourcegraph_7ab8a38d-e78b-4dc9-854b-01050704b1fb_1": name "sg-executor-job-container_sg-executor-job-codeintel-352-step.kubernetes.pre-index.0-x72q9_sourcegraph_7ab8a38d-e78b-4dc9-854b-01050704b1fb_1" is reserved for "33daf3153aa995df234b29ee7cd54978262679e7baf1551f066c0257600f0019"
   ```

### General
- Canceling a batch change execution doesn't seem to clean up the jobs (although now I think of it, it might be related to `KUBERNETES_KEEP_JOBS` set to `true`). In some cases, I saw the following error:
   ```shell
   1682940845515412670	INFO	executor_processor	workerutil/worker.go:184	github.com/sourcegraph/sourcegraph/internal/workerutil.(*Worker[...]).Start.func1	Found jobs to cancel	{"Resource": {"service.name": "executor", "service.version": "0.0.0+dev", "service.instance.id": "sg-executor-batches-6b6c7566c6-6svss"}, "Attributes": {"name": "executor_batches_worker", "IDs": [232, 226, 224, 230, 228, 229, 225, 233, 227, 231]}}
   1682940845553928348	WARN	executor_processor.Handle	workerutil/worker.go:406	github.com/sourcegraph/sourcegraph/internal/workerutil.(*Worker[...]).handle	Marked record as failed	{"Resource": {"service.name": "executor", "service.version": "0.0.0+dev", "service.instance.id": "sg-executor-batches-6b6c7566c6-6svss"}, "Attributes": {"error": "running command \"step.kubernetes.step.0.pre\": waiting for job to complete: retrieving job: client rate limiter Wait returned an error: context canceled"}}
   ```

- When a job fails, the log from the pod is not surfaced in the UI. Example:
   ![Screenshot 2023-05-02 at 12 21 54](https://user-images.githubusercontent.com/2979513/236015260-8edaabb8-b363-459e-9180-7d78d4b0c5db.png)

   Pod log:
   ```shell
   + '[' '' ']'
   + echo 'No netrc config set, continuing'
   + go mod download
   No netrc config set, continuing
   verifying github.com/docker/docker@v0.0.0-20170502054910-90d35abf7b35: checksum mismatch
   	downloaded: h1:ly3dRUfvdP5i/t9iqVHd2VQQIDtO3tpfFWPah7g4CFw=
   	go.sum:     h1:zQuy/ry/KtSvCczhEPo+ud47S8alruA/Z9NDlz7EzVo=
   
   SECURITY ERROR
   This download does NOT match an earlier download recorded in go.sum.
   The bits may have been replaced on the origin server, or an attacker may
   have intercepted the download attempt.
   
   For more information, see 'go help module-auth'.
   ```


### (Draft) PRs:
Here are PRs that make a start on our deployment methods:
- https://github.com/sourcegraph/deploy-sourcegraph-k8s/pull/26
- https://github.com/sourcegraph/deploy-sourcegraph-helm/pull/279

### Slightly related
- The `dind` executor deployment implementations need some polishing (e.g. the private docker registry image has a typo, `regisry` instead of `registry`)

**Comment by Piszmog** [comment]
This is great! Thank you! Very thorough!

**Comment by Piszmog** [comment]
- [ ] Configurable security context on job (Merge #51418)
- [ ] Ensure jobs are cleaned up when a batch change is cancelled
- [ ] Update the development/example K8s manifest with the `EXECUTOR_KUBERNETES_NODE_NAME` change
- [ ] Add information about consequences on using `EXECUTOR_KUBERNETES_NODE_NAME`
- [ ] Create Issue to followup on Metrics
- [ ] Document `externalUrl`

**Comment by sanderginn** [comment]
Testing the improvements/papercut PR: looks good, both batches and codeintel are doing its thang 🎉 

Last items:
- Add option to set the following on generated jobs:
  - Node tolerations when customers use node taints
  - Pod affinity/anti-affinity
- Due to the fact that jobs get scheduled on a single node, I eventually ran into an `OutOfPods` error, as Kubernetes has a max of 110 pods per node (disclaimer: I don't know if this rings true for _all_ distros of k8s out there, but it at least applies to vanilla and GKE). It would be good to add a `>NOTE:` block in the docs to lower `EXECUTOR_MAXIMUM_NUM_JOBS` when this occurs.

**Comment by Piszmog** [comment]
Closing since work has completed
